### DISCUSSION TITLE
Discrepancy between default file associations and program called by Fsearch

### DISCUSSION THREAD

**Discussion by pglpm** [cboxdoerfer/fsearch#439] [Q&A] [answered]
I am on Ubuntu 20.04 with KDE Plasma 5.18.8, Fsearch 0.3.alpha0.

The output of `xdg-mime query default audio/mp3` is `audiompv.desktop`, and indeed if I have some audio file "somefile.mp3" and issue the command `xdg-open somefile.mp3` or double-click on it in a dolphin folder, the file is opened with "audiompv.desktop" (which is a wrapper for MPV).

However, if I search for that file in Fsearch and once found I double click on it, it is opened by VLC.

So I wanted to ask: which file/mime-association configuration file is Fsearch using? That way I can modify it and fix this little problem.

Cheers!

PS: I don't consider this a "bug" because I'm on an originally Ubuntu/Gnome system on which I installed KDE. So I expect some glitches with file associations, and don't think it's Fsearch's fault.

**Comment by cboxdoerfer** [comment, marked as answer]
Hi,

FSearch is using the standard freedesktop mime configuration: https://specifications.freedesktop.org/mime-apps-spec/mime-apps-spec-latest.html

So your user-overrides for default applications will be stored in `~/.config/mimeapps.list`

One quick way to change the default to audiompv is with: `gio mime audio/mp3 audiompv.desktop`

**Comment by pglpm** [comment]
Thank you. It seems my `$XDG_CONFIG_HOME` is unset, and Fsearch ends up using some snap configuration directories... I'll try to set `$XDG_CONFIG_HOME` and see what happens.

**Reply by skjerns** [reply]
if `$XDG_CONFIG_HOME` is undefined, it should just default to `$HOME/.config` https://superuser.com/a/425712/532538

**Comment by pglpm** [comment]
By the way, no my `~/.config/mimeapps.list` already has the correct entry. I suspect that one of the many "mimeapps.list" in the snap subdirectories is being read instead.

**Reply by cboxdoerfer** [reply]
You could try the following command to find the mimeapps.list files which override `audio/mp3` and quickly see what default applications they set: `find ~ -type f -name "*mimeapps.list" -print0 | xargs -0 grep "audio/mp3"`

**Reply by pglpm** [reply]
Thank you for the tip! I tried it and strangely I do get the correct associations:
`/home/[myhomefolder]/.config/mimeapps.list:audio/mp3=audiompv.desktop;vlc_vlc.desktop;
/home/[myhomefolder]/.config/mimeapps.list:audio/mp3=audiompv.desktop`.

I'll try setting `$XDG_CONFIG_HOME` and report what happens. Can't do it immediately as apparently I need to log out to make it work, but have some work running at the moment.

**Reply by cboxdoerfer** [reply]
What happens when you launch `gio open /path/to/some/mp3/file.mp3` with a valid mp3 file? And what is the output of `gio mime audio/mp3` ?

**Reply by pglpm** [reply]
`gio open somefile.mp3` opens that file with VLC (the undesired behaviour).

```
$ gio mime audio/mp3
Default application for “audio/mp3”: vlc_vlc.desktop
Registered applications:
        vlc_vlc.desktop
        deadbeef.desktop
        mpv.desktop
        org.gnome.Totem.desktop
Recommended applications:
        vlc_vlc.desktop
        deadbeef.desktop
        mpv.desktop
        org.gnome.Totem.desktop
```

**Reply by pglpm** [reply]
However, `/etc/gnome/defaults.list` has the correct association. This is really baffling... where is `gio` taking that from?

But this is really a problem of my system, not or Fsearch (unless you plan of making Fsearch less dependent on this possible setting variability and glitches).

**Reply by cboxdoerfer** [reply]
Ok, so that means there's likely an issue on your system. What's interesting is that `gio` doesn't even list `audiompv.desktop` as a registered application to handle `audio/mp3`. This either means that it can't find the desktop file or that the desktop file is somehow invalid, e.g. it has syntax issues which KDE ignores but GIO doesn't.

So the first thing I suggest is to check if the desktop file is fine with `desktop-file-validate /path/to/audiompv.desktop` — this utility is part of the `desktop-file-utils` package.

**Reply by pglpm** [reply]
OK, it seems the mp3-VLC association appears in `"x-cinnamon-mimeapps.list`; I can't find any other `.list` files where it appears. Strange because I don't remember ever installing cinnamon. I'll try try renaming that file to make it useless and see what happens.

**Reply by pglpm** [reply]
Thank you Christian! `desktop-file-validate` (which I didn't know about) found two errors in the desktop file. Corrected them, now works like a charm :) 

Thank you very much for this.

**Reply by cboxdoerfer** [reply]
Great! No problem.

**Comment by pglpm** [comment]
Sorry to reopen this thread. The strange behaviour of using a non-default app upon double-clicking is happening again, with most file types. One example is `.png` files, which have `org.kde.gwenview.desktop` as default, but are opened by Fsearch with ImageMagick.

I did all checks above and can see that `org.kde.gwenview.desktop` is both system's and user's default (also done `gio mime` and even `sudo gio mime`, just to be sure), and moreover its desktop file is not corrupted (`desktop-file-validate`). This leads me to two slightly unrelated questions:

1. Has something changed in Fsearch since October 2022 that may be causing this?
2. If I experiment by doing system changes about mime types, is it enough to restart Fsearch for it to see the changes, or do I have to log out and in first?

Cheers!

**Reply by cboxdoerfer** [reply]
Hi, no problem.

> Has something changed in Fsearch since October 2022 that may be causing this?

Not really.

> If I experiment by doing system changes about mime types, is it enough to restart Fsearch for it to see the changes, or do I have to log out and in first?

Usually you don't even have to restart FSearch at all, because the changes should be picked up immediately. But you can restart FSearch, just to be sure. Logging in/out is not necessary.

Can you do the following?

1. Paste the output of `gio mime image/png`
2. Run `gio mime image/png org.kde.gwenview.desktop`
3. Again post the output of `gio mime image/png`
4. Tell me which applications opens when you run `gio open $PATH_TO_SOME_PNG_FILE`

**Comment by pglpm** [comment]
Thank you again for looking into this!

```
~$ gio mime image/png
Default application for “image/png”: org.kde.gwenview.desktop
Registered applications:
        org.kde.gwenview.desktop
        XnView.desktop
        okularApplication_kimgio.desktop
        gimp.desktop
        display-im6.q16.desktop
        wine-extension-png.desktop
        firefox.desktop
        org.kde.kolourpaint.desktop
        com.github.flxzt.rnote.desktop
Recommended applications:
        org.kde.gwenview.desktop
        XnView.desktop
        okularApplication_kimgio.desktop
        gimp.desktop
        display-im6.q16.desktop
        wine-extension-png.desktop
        firefox.desktop
        org.kde.kolourpaint.desktop
        com.github.flxzt.rnote.desktop
~$ 
~$ gio mime image/png org.kde.gwenview.desktop
Set org.kde.gwenview.desktop as the default for image/png
~$ 
~$ gio mime image/png
Default application for “image/png”: org.kde.gwenview.desktop
Registered applications:
        XnView.desktop
        okularApplication_kimgio.desktop
        emacsclient_dupe.desktop
        gimp.desktop
        display-im6.q16.desktop
        wine-extension-png.desktop
        firefox.desktop
        org.kde.kolourpaint.desktop
        com.github.flxzt.rnote.desktop
        org.kde.gwenview.desktop
Recommended applications:
        XnView.desktop
        okularApplication_kimgio.desktop
        emacsclient_dupe.desktop
        gimp.desktop
        display-im6.q16.desktop
        wine-extension-png.desktop
        firefox.desktop
        org.kde.kolourpaint.desktop
        com.github.flxzt.rnote.desktop
        org.kde.gwenview.desktop
```

Then I tried `gio open /usr/share/icons/hicolor/256x256/apps/thunderbird.png` and it does open in Gwenview (`org.kde.gwenview.desktop`).

If it's important, let me add that I (still) use KDE/Plasma as desktop, but `xdg-open` still invokes Gwenview, and my KDE config doesn't have specially configured mime types – but I'm looking into this latter point.

**Comment by cboxdoerfer** [comment]
Thx, that's weird, because everything looks and works fine here and FSearch runs the same code as `gio open`. What's the output of `fsearch --version` and `whereis fsearch`?

**Comment by pglpm** [comment]
```
~$ fsearch --version
FSearch 0.3.alpha0 (PPA-nightly)
~$ 
~$ whereis fsearch
fsearch: /usr/bin/fsearch /usr/share/man/man1/fsearch.1.gz
```

I strongly suspect that there's something wrong in my system config, not in Fsearch, and am looking into this. That's why I asked if Fsearch had relevant changes about this.

**Reply by cboxdoerfer** [reply]
Does the issue also happen you manually run FSearch from the command line (`/usr/bin/fsearch`), instead of launching it from the system menu/panel/dock?

**Comment by pglpm** [comment]
**Update**: it **is** opening in Gwenview now, upon double-click!

Edit: also other file types for which I had similar problems are now behaving fine. I really wonder which of the commands above made the relevant change???

Edit 2: I removed some .desktop files that were corrupted – although unrelated to images. I can't think of anything else that may have led to this change now. Perhaps the system scans through all desktop files before choosing the right app, and corrupted desktop files break this? I don't really understand how these things work, unfortunately.

**Reply by cboxdoerfer** [reply]
That's strange, according to `gio mime` Gwenview has been the default all the time. Do you by chance have the FSearch Snap package installed? Maybe you were using it by accident. With `snap list` you can get a list of installed snaps.

**Reply by pglpm** [reply]
No, I made sure to install from `http://ppa.launchpad.net/christian-boxdoerfer/fsearch-daily/ubuntu` and that there aren't any possible snaps (not a fan of snaps).

**Reply by pglpm** [reply]
Sorry for having stolen your time with something that isn't Fsearch's fault, Christian! And thank you for your always kind availability.

**Reply by pglpm** [reply]
After the last `gio mime image/png` I removed an `emacs_dupe.desktop` which was corrupted, and appeared in the Registered and Recommended applications. I can't think of anything else that may have corrected the app launch.

**Reply by cboxdoerfer** [reply]
No problem, I'm glad it's working now.

**Comment by tmbk1** [comment]
I was having the same problem with FSearch on KDE neon / KDE Plasma 5.27: When double-clicking results in FSearch, some files would not open in the default applications associated with their files. E.g., PDF files are normally opened in Okular, JPG files in Gwenview, but from within FSearch both would open in Vivaldi web browser. Per the above, I checked  `mimeapps.list`, but the only occurrence of "PDF" in there is in the line `application/pdf=okularApplication_pdf.desktop;`, which should have been fine.

I then checked KDE Plasma's System Settings > File Associations and found that the Application Preference Order for PDF files had Okular at the top and Vivaldi at the bottom (!) of the list. For JPG files, Gwenview was at the top, Vivaldi near the bottom of the list. I have no idea what made FSearch open files in Vivaldi, out of all entries on these ordered lists. 

However, after manually deleting Vivaldi from the Application Preference Order for PDF and JPG files, I found that double-clicking them in FSearch resulted in the expected behaviour, i.e. opening in Okular and Gwenview, respectively. 

I'm leaving this here, perhaps others will find it helpful.

**Reply by skjerns** [reply]
Thanks! that actually finally fixed it for me. For me it took the second to last entry on some things, strange.
